librados: modify Pipe::connect() to return the error code - part 1

vumrao opened the pull request:
librados: modify Pipe::connect() to return the error code - part 1
Fixes: http://tracker.ceph.com/issues/15308
Signed-off-by: Vikhyat Umrao <vumrao@redhat.com>

vumrao commented:
@jdurgin Hello Josh as discussed I have sent this first part of fix for pipe::connect() for tracker : http://tracker.ceph.com/issues/15308

gregsfortytwo commented:
You can't just set errno on all those callchains. tcp_read() calls into tcp_read_wait() which returns -1 if the socket has been closed, for instance — nothing to do with errno.

I am also not 100% sure that you can rely on errno's value being set correctly once you've exited your initial function context; the rules for it around that or threading or something are a bit funky.

vumrao commented:
Thank you Greg for your inputs.

>You can't just set errno on all those callchains. tcp_read() calls into tcp_read_wait() which returns -1 if the socket has been closed, for instance — nothing to do with errno.

Sure I will make the change for tcp_read and then I checked Pipe::do_sendmsg() and same goes for do_sendmsg(). This function do have a standard call to sendmsg() and I think we can handle errno properly here. I am sending the next patch set with making changes to do_sendmsg().
Please let me know your inputs.

> I am also not 100% sure that you can rely on errno's value being set correctly once you've exited your initial function context; the rules for it around that or threading or something are a bit funky.

For this I had discussion with Josh and because of this I made it as part 1 . We discussed any developer who has more background of code , he can send patch for part 2. Part 2 patch will take pipe::connect() returned errno to reach properly to rados_connect().

yuyuyu101 commented:
I think part 2 is a little difficult for current impl. we don't have any way to notify the caller who want to connect to a endpoint except ms_handle*.

gregsfortytwo commented:
>>I am also not 100% sure that you can rely on errno's value being set correctly once you've exited your initial function context; the rules for it around that or threading or something are a bit funky.

>For this I had discussion with Josh and because of this I made it as part 1 . We discussed any developer who has more background of code , he can send patch for part 2. Part 2 patch will take pipe::connect() returned errno to reach properly to rados_connect().

I mean, you may need to return errno from do_sendmsg and tcp_read, rather than reading it out in Pipe::connect directly. Not certain, though.

vumrao commented:
> I think part 2 is a little difficult for current impl. we don't have any way to notify the caller who want to connect to a endpoint except ms_handle*. 

Thanks @yuyuyu101 for your inputs. I am sure it is little difficult :). @jdurgin and I discussed the same.

vumrao commented:
> I mean, you may need to return errno from do_sendmsg and tcp_read, rather than reading it out in Pipe::connect directly. Not certain, though.

Thanks Greg, sorry I misunderstood it. In `do_sendmsg()` I had added `errno`. Now I have added `errno` in `tcp_read` , `tcp_read_wait()` and `tcp_read_nonblocking()`.

jdurgin commented on src/msg/simple/Pipe.cc:
```diff
@@ -2440,8 +2453,11 @@ int Pipe::write_message(const ceph_msg_header& header, const ceph_msg_footer& fo
-    return -1;
+  int rc = 0;
+  if (sd < 0) {
+    rc = -errno;
```
errno isn't getting set by anything - it'll just have an unrelated value here.

For these sanity checks, tcp_{read,write} should return a specific error, perhaps EINVAL

jdurgin commented on src/msg/simple/Pipe.cc:
```diff
@@ -2452,25 +2468,32 @@ int Pipe::tcp_read(char *buf, unsigned len)
-    if (tcp_read_wait() < 0)
-      return -1;
+    if (tcp_read_wait() < 0) {
+      rc = -errno;
```
reading errno should be moved to right after the syscall, not in any of the callers of these wrappers. that'll avoid as much as possible getting invalid errno values, and avoiding accidental resetting of it if the code changes

jdurgin commented on src/msg/simple/Pipe.cc:
```diff
@@ -2565,18 +2594,20 @@ ssize_t Pipe::buffered_recv(char *buf, size_t len, int flags)
 ssize_t Pipe::tcp_read_nonblocking(char *buf, unsigned len)
 {
+  int rc = -1;
   ssize_t got = buffered_recv(buf, len, MSG_DONTWAIT );
```
same thing here, the errno value is only useful all the way in do_recv here, buffered_recv and tcp_read_nonblocking should just pass the return code up, and not read errno themselves, since they are not calling any syscalls that set it

vumrao commented:
@jdurgin Thank you Josh for your inputs and explaining things in detail. Sending new patch set.

jdurgin commented:
lgtm now, thanks

vumrao commented:
Thank you Josh!

tchaikov commented on src/msg/simple/Pipe.cc:
```diff
@@ -901,7 +901,8 @@ int Pipe::connect()
   if (sd < 0) {
-    lderr(msgr->cct) << "connect couldn't created socket " << cpp_strerror(errno) << dendl;
+    rc = -errno;
+    lderr(msgr->cct) << "connect couldn't created socket " << cpp_strerror(rc) << dendl;
```
why change this line, `cpp_strerror(errno)` is better here.

vumrao replied:
Thanks Kefu, I have made the change.

dzafman replied:
I prefer to keep this change, since it is better to get errno saved before something might make another system call before cpp_strerror() is called now or in the future.

tchaikov replied:
thanks @dzafman, josh also reminded me this at https://github.com/ceph/ceph/pull/8416#discussion_r58825773 .

tchaikov commented on src/msg/simple/Pipe.cc:
```diff
@@ -2244,7 +2257,7 @@ int Pipe::do_sendmsg(struct msghdr *msg, unsigned len, bool more)
   }
   restore_sigpipe();
-  return 0;
+  return rc;
```
why you are changing this line? imo, it would be easier for the reader if we are sure that this branch always return `0`?

tchaikov commented on src/msg/simple/Pipe.cc:
```diff
@@ -2495,7 +2511,7 @@ int Pipe::tcp_read_wait()
     return -1;
 
-  return 0;
+  return rc;
```
ditto.

tchaikov commented on src/msg/simple/Pipe.cc:
```diff
@@ -2210,16 +2221,18 @@ int Pipe::do_sendmsg(struct msghdr *msg, unsigned len, bool more)
+      ldout(msgr->cct,1) << "do_sendmsg error " << cpp_strerror(rc) << dendl;
       restore_sigpipe();
-      return -1;
+      return rc;
```
could return -errno right away without introducing a local "rc" here.

vumrao replied:
@tchaikov ack will make the changes.

jdurgin commented on src/msg/simple/Pipe.cc:
```diff
@@ -901,6 +901,7 @@ int Pipe::connect()
   sd = ::socket(peer_addr.get_family(), SOCK_STREAM, 0);
   if (sd < 0) {
+    rc = -errno;
     lderr(msgr->cct) << "connect couldn't created socket " << cpp_strerror(errno) << dendl;
```
the reason to avoid cpp_strerror(errno) here is perhaps a bit paranoid, but it's to avoid errno being changed before we get to printing it here.

e.g. lderr() or operator << or other things called indirectly could call some function that sets errno, causing us to lose the original error code. hence the preference for cpp_strerror(rc).

vumrao replied:
Thanks Josh for your inputs, sorry I missed to explain this to Kefu, why we did like this.

tchaikov replied:
josh, makes sense. thanks for the explanation. yeah, `clock_gettime(2)` could set errno.

tchaikov commented:
lgtm

vumrao commented:
Thank you Josh and Kefu for your help in this PR.

vumrao commented:
doing rebase to trigger the build again.

vumrao commented:
@liewegas Thank you.
